README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# scatterwithoutlier

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
<!-- badges: end -->

The goal of scatterwithoutlier is to create scatterplot for two
quantitative variables with labeled outliers. This function takes a
dataset(as a list, and dataframes/tibbles are lists) of two quantitative
variables and fits a single linear regression model from them, then it
displays a scatterplot with fitted regression line. Also, it displays
any influential points as red color based on user defined criteria
(cut-off for cook’s distance).

## Development Process

Overall, this package was developed using the modern functions usethis()
and devtools(), with some manual edits. Specifically:

usethis or devtools functions applied:

  - create\_package(“scatterwithoutlier”) and use\_git() applied to
    generate the local package.
  - use\_r(“scatterwithoutlier”) generated scatterwithoutlier.R in the R
    folder.
  - use\_testthat() generated the tests folder and associated files.
  - use\_test(“scatterwithoutlier”) generated test-scatterwithoutlier.R
    in tests folder.
  - use\_package(“dplyr”) generated the dependecy on package “dplyr”.
  - use\_pipe() generated the dependecy on package “magrittr”.
  - use\_package(“gapminder”) generated the dependecy on package
    “gapminder”.
  - use\_package(“stats”) generated the dependecy on package “stats”.
  - use\_readme\_rmd() generated README.Rmd.
  - build\_readme() generated README.md from README.Rmd.
  - use\_vignette(“scatterwithoutlier”) generated vignettes.
  - build\_vignettes() built the vignettes.
  - use\_mit\_license(“Tianyi Zheng”) generated LICENSE.md.
  - use\_code\_of\_conduct() generated CODE\_OF\_CONDUCT.md.
  - usethis::use\_pkgdown() and pkgdown::build\_site() generated the
    package website.

Manual works:

  - manually created a project on Github website, and pushed the local
    package into the repo.
  - manually implemented scatterwithoutlier.R and
    test-scatterwithoutlier.R based on Assignment 1.
  - manually filled in information for Readme, and vignettes.
  - manually created and filled in information for NEWS.md.
  - manually added a badge.

Also, tests are developed to validate the function is outputing a ggplot
object, and raising an error upon unexpected inputs.

## Installation

`scatterwithoutlier` is not on CRAN yet, so the best way to install it
is:

``` r
# install.packages("devtools")
devtools::install_github("tianyica/scatterwithoutlier")
```

## Example

``` r
library(scatterwithoutlier)
```

The following example plots the relationship between life expectancy (as
independent variable) and population (as dependent variable) from the
gapminder dataset on the default settings.

``` r
scatterwithoutlier(gapminder::gapminder %>%
     dplyr::select(lifeExp,pop))
#> `geom_smooth()` using formula 'y ~ x'
```

<img src="man/figures/README-unnamed-chunk-2-1.png" width="100%" />

The following example plots the relationship between life expectancy (as
independent variable) and population (as dependent variable) from the
gapminder dataset on the threshold of 0.05 on cook’s distance, and
outputing the message on the count of influential observations and the
process of the function.

``` r
scatterwithoutlier(gapminder::gapminder %>%
     dplyr::select(pop,lifeExp),xy=FALSE,cooksd=0.05,verbose = TRUE,outputcount = TRUE)
#> Linear model fitting done, now creating plot...We have 4 influential observations at 0.05 cook's distance threshold
#> `geom_smooth()` using formula 'y ~ x'
```

<img src="man/figures/README-unnamed-chunk-3-1.png" width="100%" />

The following example plots the relationship between life expectancy (as
independent variable) and GDP per capita (as dependent variable) from
the gapminder dataset on the threshold of 0.05 on cook’s distance.

``` r
scatterwithoutlier(gapminder::gapminder %>%
     dplyr::select(lifeExp,gdpPercap),cooksd = 0.05)
#> `geom_smooth()` using formula 'y ~ x'
```

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" />


NEWS.md
# scatterwithoutlier 0.1.0

The first version of `scatterwithoutlier` is now available! Its functionality is rather limited at the moment, but is still useful. The `scatterwithoutlier` function takes a dataset(as a list, and dataframes/tibbles are lists) of two quantitative variables and fits a single linear regression model from them, then it displays a scatterplot with fitted regression line. Also, it displays any influential points as red color based on user defined criteria (cut-off for cook’s distance).



tests/testthat.R
library(testthat)
library(scatterwithoutlier)

test_check("scatterwithoutlier")


tests/testthat/test-scatterwithoutlier.R
#Testing that the function would produce error on unexpected inputs.
test_that("Unexpected Input Testing",{
  expect_error(scatterwithoutlier(data=gapminder::gapminder %>%
                                    dplyr::select(country,pop)))
  expect_error(scatterwithoutlier(data=gapminder::gapminder %>%
                                    dplyr::select(gdpPercap,pop),xy=1))
  expect_error(scatterwithoutlier(data=gapminder::gapminder %>%
                                    dplyr::select(gdpPercap,pop),verbose ="TRUE"))
  expect_error(scatterwithoutlier(data=gapminder::gapminder %>%
                                    dplyr::select(gdpPercap,pop),outputcount = 0))
  expect_error(scatterwithoutlier(data=gapminder::gapminder %>%
                                    dplyr::select(year,country,pop)))
  expect_error(scatterwithoutlier(data=FALSE))
  expect_error(scatterwithoutlier(data=gapminder::gapminder %>%
                                    dplyr::select(gdpPercap,pop),cooksd = "0.5"))

})

#Testing the output of the function is a ggplot object.
test_that("Output check",{
  expect_equal(class(scatterwithoutlier(data=gapminder::gapminder %>%
                                          dplyr::select(gdpPercap,pop))),c("gg","ggplot"))
  expect_output(scatterwithoutlier(data=gapminder::gapminder %>%
                                     dplyr::select(gdpPercap,pop),outputcount = TRUE))
})


R/scatterwithoutlier.R
#' Create scatterplot for two quantitative variables with labeled outliers
#'
#' @param data a list (dataframe, tibble, etc…) with two quantitative variables.
#' @param xy ogicals. If TRUE then the first column of the list would assumed to be the independent variable, and vice versa for FALSE. The default if TRUE.
#' @param cooksd a non-negative numerical value for cook’s distance threshold. The default is 0.5.
#' @param outputcount logicals. If TRUE then print a message on the count of influential observations.
#' @param verbose logicals. If TRUE then print a message to inform the user that the process is running.
#' @param ... additional arguments to be passed to the low level regression fitting functions.
#'
#' @return A ggplot object(scatterplot) with fitted regression line and red labelled influential points.
#' @export
#'
#' @examples
#'
#' scatterwithoutlier(gapminder::gapminder %>%
#' dplyr::select(lifeExp,pop))
#'
#' scatterwithoutlier(gapminder::gapminder %>%
#' dplyr::select(pop,lifeExp),xy=FALSE,cooksd=0.05,verbose = TRUE,outputcount = TRUE)
#'
#' scatterwithoutlier(gapminder::gapminder %>%
#' dplyr::select(lifeExp,gdpPercap),cooksd = 0.05)

scatterwithoutlier<-function(data, xy=TRUE, cooksd=0.5, outputcount=FALSE, verbose=FALSE,...){
  if(typeof(data)!="list"){
    stop('The function expect input data as list\n',
         'but the input is: ', typeof(data)[1])
  }
  if(ncol(data)!=2){
    stop('The function expect two columns\n',
         'but the input has: ', ncol(data))
  }
  if(!is.numeric(data[[1]])|!is.numeric(data[[2]])){
    stop('The function expect quantinative/numberic variables.')
  }
  if(!is.numeric(cooksd)|cooksd<0){
    stop('cooksd needs to be a non-negative number')
  }
  if(!is.logical(xy)){
    stop('xy needs to be a boolean value')
  }
  if(!is.logical(outputcount)){
    stop('outputcount needs to be a boolean value')
  }
  if(!is.logical(verbose)){
    stop('verbose needs to be a boolean value')
  }
  if(xy==TRUE){
    x=data[[1]]
    y=data[[2]]
    xname=colnames(data)[1]
    yname=colnames(data)[2]

  }
  if(xy==FALSE){
    x=data[[2]]
    y=data[[1]]
    xname=colnames(data)[2]
    yname=colnames(data)[1]

  }
  model<-lm(y~x,...)
  if(verbose==TRUE){
    cat("Linear model fitting done, now creating plot...")
  }

  if(outputcount==TRUE){
    Infobvec<-as.vector(cooks.distance(model)>cooksd)
    Infobnum<-length(subset(Infobvec,Infobvec==TRUE))
    cat("We have",Infobnum,"influential observations","at",cooksd,"cook's distance threshold")
  }
  data %>%
    dplyr::mutate(Infob=cooks.distance(model)>cooksd) %>%
    ggplot2::ggplot(ggplot2::aes(x,y))+
    ggplot2::geom_point(ggplot2::aes(color=Infob))+
    ggplot2::geom_smooth(method=lm, se=FALSE)+
    ggplot2::scale_color_manual(values=c("Black", "Red"))+
    ggplot2::theme_bw()+
    ggplot2::labs(x=xname,y=yname)+
    ggplot2::theme(legend.position = "none")
}
